pallets/pallet-lottery/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+
+		/// Type representing the weight of this pallet
+		type WeightInfo: WeightInfo;
+
+		/// The balance type
+		type Balance: Member
+			+ Parameter
+			+ MaxEncodedLen
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Debug
+			+ From<u128>
+			+ sp_std::iter::Sum;
+
+		type RoleManager: ManageRoles<Self::AccountId>;
+
+		type Bank: BasicAccounting<Self::AccountId, Self::Balance> + GetTreasury<Self::AccountId>;
+
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = BlockNumberFor<Self>>;
+
+		type Randomness: Randomness<u32, BlockNumberFor<Self>>;
+
+		#[pallet::constant]
+		type LotteryPayoutPeriod: Get<BlockNumberFor<Self>>;
+
+		#[pallet::constant]
+		type PrizePoolAccount: Get<Self::AccountId>;
+
+		#[pallet::constant]
+		type TaxRate: Get<Percent>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The prize split total is not equal to one.
+		InvalidPrizeSplitTotal,
+		/// The account role does not equal to the expected role.
+		IncorrectRole,
+		/// The ticket price is not set.
+		TicketPriceNotSet,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Set the prize split.
+		PrizeSplitUpdated { split: Vec<Percent> },
+		/// Set the ticket price.
+		TicketPriceUpdated { old: T::Balance, new: T::Balance },
+		/// Customer bought numbers of tickets.
+		TicketsBought { id: T::AccountId, number: u32 },
+		/// Customer won the lottery fund and paid tax.
+		LotteryWon { user: T::AccountId, won_fund: T::Balance, tax: T::Balance },
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn prize_split)]
+	pub type PrizeSplit<T: Config> = StorageValue<_, Vec<Percent>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn ticket_price)]
+	pub type TicketPrice<T: Config> = StorageValue<_, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn tickets)]
+	pub type PlayersAndLotteries<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, u32, ValueQuery>;
+
+	#[pallet::storage]
+	pub(crate) type RandomSeed<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::genesis_config]
+	#[derive(frame_support::DefaultNoBound)]
+	pub struct GenesisConfig<T: Config> {
+		phantom: PhantomData<T>,
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+		fn build(&self) {
+			// Set PrizeSplit storage with Percent::one()
+			PrizeSplit::<T>::put(vec![Percent::one()]);
+		}
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn integrity_test() {}
+
+		fn on_finalize(block_number: BlockNumberFor<T>) {
+			// check if we should payout this block
+			if (block_number % T::LotteryPayoutPeriod::get()).is_zero() {
+				if let Ok(treasury) = T::Bank::treasury() {
+					// Set up data for choosing winners
+					let total = T::Bank::free_balance(&T::PrizePoolAccount::get());
```

Comment:
This can be moved to a separate function (resolve_lottery_winner) to kee on_finalize clean